src/emulate.c

```diff
@@ -1407,6 +1449,92 @@ static void extend_block(riscv_t *rv, block_t *block)
         last_ir->branch_untaken = next->ir;
 }
 
+static void match_pattern(block_t *block)
+{
+    for (uint32_t i = 0; i < block->n_insn - 1; i++) {
+        rv_insn_t *ir = block->ir + i, *next_ir = NULL;
+        int32_t count = 0;
+        switch (ir->opcode) {
+        case rv_insn_auipc:
+            next_ir = ir + 1;
+            if (next_ir->opcode == rv_insn_addi) {
+                if (ir->rd == next_ir->rs1) {
+                    ir->opcode = rv_insn_fuse1;
+                    ir->rd = next_ir->rd;
+                    ir->imm2 = next_ir->imm;
+                    ir->impl = dispatch_table[ir->opcode];
+                    next_ir->opcode = rv_insn_empty;
+                    next_ir->impl = dispatch_table[next_ir->opcode];
+                } else if (ir->rd == next_ir->rs2) {
+                    ir->opcode = rv_insn_fuse2;
+                    ir->rd = next_ir->rd;
+                    ir->rs1 = next_ir->rs1;
+                    ir->impl = dispatch_table[ir->opcode];
+                    next_ir->opcode = rv_insn_empty;
+                    next_ir->impl = dispatch_table[next_ir->opcode];
+                }
+            }
+            break;
+        case rv_insn_sw:
+            count = 1;
+            for (uint32_t j = 1; j < block->n_insn - 1 - i; j++) {
+                next_ir = ir + j;
+                if (next_ir->opcode != rv_insn_sw)
+                    break;
+                count++;
+            }
+            if (count >= 5) {
+                ir->opcode = rv_insn_fuse3;
+                ir->mem_fuse = malloc(count * sizeof(mem_fuse_t));
+                ir->imm2 = count;
+                ir->mem_fuse[0].imm = ir->imm;
+                ir->mem_fuse[0].rd = ir->rd;
+                ir->mem_fuse[0].rs1 = ir->rs1;
+                ir->mem_fuse[0].rs2 = ir->rs2;
+                ir->impl = dispatch_table[ir->opcode];
+                for (int j = 1; j < count; j++) {
+                    next_ir = ir + j;
+                    ir->mem_fuse[j].imm = next_ir->imm;
```

Comment:
Instead of various assignments, can you utilize `memcpy` for structure member filling? `memcpy` usually outperforms.